admiral/pkg/clusters/handler.go

```diff
@@ -323,3 +325,94 @@ func GetServiceWithSuffixMatch(suffix string, services []*coreV1.Service) string
 	}
 	return ""
 }
+
+func HandleEventForClientDiscovery(ctx context.Context, event admiral.EventType, obj *common.K8sObject,
+	remoteRegistry *RemoteRegistry, clusterName string) error {
+	log.Infof(LogFormat, event, obj.Type, obj.Name, clusterName, common.ReceivedStatus)
+	globalIdentifier := common.GetGlobalIdentifier(obj.Annotations, obj.Labels)
+	originalIdentifier := common.GetOriginalIdentifier(obj.Annotations, obj.Labels)
+	if len(globalIdentifier) == 0 {
+		log.Infof(LogFormat, event, obj.Type, obj.Name, clusterName, "Skipped as '"+common.GetWorkloadIdentifier()+" was not found', namespace="+obj.Namespace)
+		return nil
+	}
+	ctxLogger := common.GetCtxLogger(ctx, globalIdentifier, "")
+	if commonUtil.IsAdmiralReadOnly() {
+		ctxLogger.Infof(common.CtxLogFormat, event, "", "", "", "processing skipped as Admiral is in Read-only mode")
+		return nil
+	}
+
+	// Should not return early here for TrafficConfig persona, as cache should build up during warm up time
+	if IsCacheWarmupTime(remoteRegistry) && !common.IsPersonaTrafficConfig() {
+		ctxLogger.Infof(common.CtxLogFormat, event, "", "", "", "processing skipped during cache warm up state")
+		return fmt.Errorf(common.CtxLogFormat, event, "", globalIdentifier, "", "processing skipped during cache warm up state for env="+" identity="+globalIdentifier)
+	}
+
+	//if we have a deployment/rollout in this namespace skip processing to save some cycles
+	if DeploymentOrRolloutExistsInNamespace(remoteRegistry, globalIdentifier, clusterName, obj.Namespace) {
+		log.Infof(LogFormatAdv, "Process", obj.Type, obj.Name, obj.Namespace, clusterName, "Skipping client discovery as Deployment/Rollout already present in namespace for client="+globalIdentifier)
+		return nil
+	}
+
+	ctx = context.WithValue(ctx, "clusterName", clusterName)
+	ctx = context.WithValue(ctx, "eventResourceType", obj.Type)
+
+	if remoteRegistry.AdmiralCache != nil {
+
+		UpdateIdentityClusterCache(remoteRegistry, globalIdentifier, clusterName)
+
```

Comment:
I think we should return an error if `remoteRegistry.AdmiralCache == nil` and not move ahead because if `UpdateIdentityClusterCache` is not called then calling Added of dependency controller might not do anything.

Reply:
Good catch!